rllib/utils/replay_buffers/mixin_multi_agent_replay_buffer.py

```diff
@@ -0,0 +1,293 @@
+            If not enough new samples are available, mixes in less old samples
+            to retain self.replay_ratio on average.
+            """
+            def round_up_or_down(value, ratio):
+                """Returns an integer averaging to value*ratio."""
+                product = value * ratio
+                ceil_prob = product % 1
+                if random.uniform(0, 1) < ceil_prob:
+                    return int(np.ceil(product))
+                else:
+                    return int(np.floor(product))
+
+            max_num_new = round_up_or_down(num_items, 1 - self.replay_ratio)
+            # if num_samples * self.replay_ratio is not not round,
+            # we need one more sample with a probability of
+            # (num_items*self.replay_ratio)%1
+
+            _buffer = self.replay_buffers[_policy_id]
+            output_batches = self.last_added_batches[_policy_id][:max_num_new]
+            self.last_added_batches[_policy_id] = self.last_added_batches[
+                                                      _policy_id][max_num_new:]
+
+            # No replay desired
+            if self.replay_ratio == 0.0:
+                return SampleBatch.concat_samples(output_batches)
+            # Only replay desired
+            elif self.replay_ratio == 1.0:
+                return _buffer.sample(num_items,
+                                      beta=self.prioritized_replay_beta)
+
+            if np.isclose(max_num_new, num_items * (1 - self.replay_ratio)):
+                # The optimal case, we can mix in a round number of old
+                # samples on average
+                num_old = num_items - max_num_new
+            else:
+                num_new = len(output_batches)
+                # We never want to return more elements than num_items
+                num_old = min(num_items - max_num_new,
+                              round_up_or_down(num_new,
+                                               self.replay_ratio
+                                               / (1 - self.replay_ratio)))
+
+            output_batches.append(_buffer.sample(num_old,
+                                                 self.prioritized_replay_beta))
+            return SampleBatch.concat_samples(output_batches)
+
+        def check_buffer_is_ready(_policy_id):
+            if (len(self.replay_buffers[policy_id]) == 0) or \
+                (len(self.last_added_batches[_policy_id]) == 0 and
+                 self.replay_ratio < 1.0):
+                return False
+            return True
+
+        with self.replay_timer:
+            if self.replay_mode == "lockstep":
+                assert (
+                    policy_id is None
+                ), "`policy_id` specifier not allowed in `locksetp` mode!"
+                if check_buffer_is_ready(_ALL_POLICIES):
+                    return mix_batches(_ALL_POLICIES)
+            elif policy_id is not None:
+                if check_buffer_is_ready(policy_id):
```

Comment:
don't quite get this logic, if buffer is not ready, we silently return None?

Reply:
My bad, this should return an empty MultiAgentBatch I think!